### PR TITLE
Hw16

### PR DESCRIPTION
**Pull request by VolodiaH**

### CODE REVIEW

File: Hw16/DynamicIntArray.cpp [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+
+	memcpy(m_data, other.m_data, m_size * sizeof(int));
+	std::fill(m_data + m_size, m_data + m_capacity, 0);
+
+	return *this;
+}
+
+int& DynamicIntArray::operator[](std::size_t index) const
+{
+	return m_data[index];
```
Comment (helen-black) [inline Hw16/DynamicIntArray.cpp:56]: what if index it out of range?

File: Hw16/DynamicIntArray.cpp [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+}
+
+void DynamicIntArray::setSize(std::size_t newSize)
+{
+	if(newSize == m_size || m_size <= 0)
```
Comment (helen-black) [inline Hw16/DynamicIntArray.cpp]: I don't understand the second condition? maybe `|| newSize < 0`

File: Hw16/DynamicIntArray.cpp [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+		return *this;
+
+	m_size = other.m_size;
+	m_capacity = other.m_capacity;
+
+	delete[]m_data;
+
+	m_data = new int[m_capacity];
+
+	memcpy(m_data, other.m_data, m_size * sizeof(int));
```
Comment (helen-black) [inline Hw16/DynamicIntArray.cpp:49]: You can use copy c-tor here

File: Hw16/DynamicIntArray.cpp [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+	
+	int *tmpArray = new int[newSize];
+
+	memcpy(tmpArray, m_data, copyCount * sizeof(int));
+
+	delete[] m_data;
+
+	m_data = tmpArray;
+
+	m_capacity = newSize;
```
Comment (helen-black) [inline Hw16/DynamicIntArray.cpp:82]: what if capacity is bigger then newSize?

File: Hw16/DynamicIntArray.cpp [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+}
+
+void DynamicIntArray::push_back(int element)
+{
+	if(m_capacity == m_size)
+	{
+		if(!m_capacity)
+			setSize(VecDefaultSize);
+		else
+			setSize(m_capacity * 1.5);
```
Comment (helen-black) [inline Hw16/DynamicIntArray.cpp]: if you have size 10 and capacity 10, you'll set size 15 and after that do operation:
```c++
m_data[m_size++] = element;
```
something is wrong here

File: Hw16/MemoryLeaksHW.cpp [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+{
+    delete[] m_str;
+    if (c_str == nullptr)
+    {
+        m_str = new char[1];
+        *m_str = '\0';
+    }
+    else
+    {
+        const std::size_t size = strlen(c_str);
```
Comment (helen-black) [inline Hw16/MemoryLeaksHW.cpp]: your size should be `strlen(c_str) + 1;` to reserve space for '\0' symbol

**Review by helen-black — CHANGES_REQUESTED** [review]

File: Hw16/DynamicIntArray.cpp (hunk 1 above)
Comment (VolodiaH) [reply Hw16/DynamicIntArray.cpp:56]: usually, operator[] does not check, ok I'll add check

File: Hw16/DynamicIntArray.cpp [hunk 7]
```diff
@@ -15,10 +16,8 @@ m_size{ 0 }
 DynamicIntArray::DynamicIntArray(std::size_t size):
 m_data{ new int[(size > 0) ? size : VecDefaultSize ] },
 m_capacity{ (size > 0) ? size : VecDefaultSize },
-m_size{ (size > 0) ? size : VecDefaultSize }
+m_size{ 0 }
```
Comment (helen-black) [inline Hw16/DynamicIntArray.cpp]: m_size should be equal to size

File: Hw16/DynamicIntArray.cpp (hunk 1 above)
Comment (helen-black) [reply Hw16/DynamicIntArray.cpp:56]: it checks in debug mode

**Review by helen-black — APPROVED** [review]

**Pull request merged**